### ISSUE TITLE
Alarms: discuss propagation and display

### ISSUE THREAD

**Issue by Tom-Willemsen** [ISISComputingGroup/IBEX#3715]
As an instrument scientist, I want to be clear about exactly what alarms on a PV mean. I want them to be displayed consistently in all the places where I might see them.

This ticket is to have a meeting to discuss how alarms should be reflected in:
- The blocks pane of the GUI
- OPIs within the GUI
- The dataweb
- The nexus file
- The data browser

The output of this meeting should be some options for the SAG to decide on.

**Comment by kjwoodsISIS** [comment]
I think we should also include what SECI currently does in our discussions, so that we know what baseline we are working from.

**Comment by FreddieAkeroyd** [comment]
SECI didn't have the concept of a separate alarm state like EPICS, so if there was a communications failure for example it would often show "NaN" as the readback to be displayed/logged.

**Comment by ChrisM-S** [comment]
Yes, the function of showing that a component of the system was "unwell" would be a count of messages in the "System Messages" side bar
![image](https://user-images.githubusercontent.com/3035405/47908521-4259d400-de85-11e8-98e7-22cef2cab58d.png)

On visiting the message panel, the errors could be shown in a message log see below.  This list of errors is analogous to the EPICS alarms view in most things (but less functional) except that they could not be cancelled/ignored in the same way.  To stop an alarm it was/(is!) necessary to stop the VI which was logging the error.
![image](https://user-images.githubusercontent.com/3035405/47908616-7c2ada80-de85-11e8-9f2b-72872ac498ad.png)

Earlier versions of SECI allowed navigation to the error(alarm) cause by navigating directly to the problem VI directly from the message which was a nice feature - I think it was a shame this was lost as it immediately took you to where the problem was without searching.  It might be worth looking at this with the EPICS alarms and OPIs.

Finally, the closest to an alarm state, in a lot of VIs, the unwell/alarm state would be shown by a VI with a "red" background while the error state persisted (whilst sending recurring alerts/errors in the "System Messages" area which would show why).

![image](https://user-images.githubusercontent.com/3035405/47910386-9d41fa00-de8a-11e8-8cc2-ce7a83fa3909.png)

**Comment by ChrisM-S** [comment]
It does occur to me that the terminology is confusing and we should be careful with this in discussions.  In common parlance "alarm" suggests a bell ringing for immediate attention which I would term an "Alert" in our context.  The EPICS alarm seems to me much more a an inventory of components which are identified to be unhappy in some way.

The distinction which is made in Nagios seems to be a good analogy, Alerts get sent to someone who needs to know something has changed but can be optionally received/triggered.  Even when not triggered the web site lists a whole bunch of unhappy things which we may or not care about.  I'm not quite sure what I would call this list, probably not alarms...

**Comment by KathrynBaker** [comment]
It is also worth noting, that on the whole, this was only comms errors that SECI reported, and it would try n times (usually 10), go into an error state and then after 30 seconds it would clear the errors and try again – so the count was easily considered as a time marker as well. Any items being read would be set as NaN or an empty string, and enums varied – usually going into a suitable value if one was included, but these were rare due to the interactions between SECI and LabVIEW.

Any error would also mean that nothing else was done until the error cleared (automatically after 30 seconds), so any calculations would become NaN or void.

High/Low errors and similar were not a concept here generally, but there are probably a few examples where we did have non-comms errors.

We do need to sort out the false alarms and consider the user experience. Chris has described the SECI interactions well.

**Comment by Tom-Willemsen** [comment]
Part of this ticket was brought up by James in reference to RIKENFE - to quote him directly:

> can we put in another request that PVs that are in an error state, and blocks that link to them, return an obviously bad value such as (numeric) NaN or (string) “Error”?

**Comment by John-Holt-Tessella** [comment]
Reply sent to discussing issues:
> Your email raises several issues, which need careful thought because they represent proposals to make fundamental changes to how IBEX works.  The essential problem as we see it is how an error in a devices communications is communicated consistently to the user such that they are aware of the error and can take appropriate action. We thought it would be helpful to share some of the discussions we have had on these issues:

> 1. On an error a block should go into an alarm state and also change its value. 
This proposal raises a couple of problems:
One problem is, if there are intermittent communications errors, then you will see the value flashing between its actual value and NaN.  Not all devices behave well and we think there will be a spectrum of views on what people want to see - some people would rather see a value, others a flashing disconnect signal. At present, a communications error is always flagged with a purple border. 
A second problem relates to the underlying PV which reports the value from the device.  PVs are part of the EPICS framework; their behaviour is determined by the design conventions that underpin EPICS.  The EPICS convention is to preserve the value of the PV and raise an error flag.  Therefore, a caget (for example) will always return the value of the PV.  We can’t change the EPICS design conventions.

> There may be ways to workaround the inherent behaviour of PVs - by changing the way they are displayed in the user interface and reported by genie_python - but these could be far-reaching.  We should explore alternative ways of providing you with the information you need.

> 2. Change waitfor so that it stalls on communication error. 
The problem here is that if you leave a script running and it hits this condition, the script stalls until a user intervenes. This is fine if someone is watching it, but if the problem occurs at at 2am, then it may not collect any data for several hours.  
The current behaviour is to continue collecting data, even though a device may have (temporarily) stopped communicating.  Even though the data is incomplete it may still be more useful than no data at all. Some instruments might find it difficult to accept such a change.
> One suggestion to address the problem is to add another parameter to waitfor so that you can choose the behaviour on an alarm/error condition. It would then be up to the script writer to choose the preferred behaviour if different from the agreed default.
> 
> BTW - essentially the same situation applies to run-control.  Run-control also ignores communication errors.  If we stall on communication errors, run-control will halt data collection until the communication error is cleared.  So, again, you could end up collecting no data for several hours if communication to a device is interrupted.
> 
> 3. Logging NaNs & INFs
Apparently this has always been the case in SECI as well. There is a limitation in the SQL-Lite software used to store data prior to its transfer to the NEXUS file.  Essentially, SQL-Lite does not handle NaNs and INFs correctly.  We have some ideas about how we could work around this limitation.

**Comment by kjwoodsISIS** [comment]
have a meeting

**Comment by KathrynBaker** [comment]
The NaN instead of 0 for devices in an error state has come up again from RIKENFE as a footprints ticket, proposing to bring it into consideration again.

**Comment by KathrynBaker** [comment]
This is likely to be superceded by the ongoing alarms work, closing.
